core/global/display_manager.gd

```diff
@@ -40,97 +31,275 @@ func supports_brightness() -> bool:
-		logger.debug("Setting brightness on " + brightness_file + " to " + str(real_value))
-		if brightness_provider == BRIGHTNESS_PROVIDER.STEAMOS:
-			if _steamos_priv_write(brightness_file, real_value) != OK:
-				return FAILED
-			continue
-		
-		return FAILED
-	
+
+	var err := brightness_provider.set_brightness(value, type, backlight)
+	if err != OK:
+		return err
 	brightness_changed.emit()
-	
-	return OK
+
+	return err
 
 
 ## Returns the current brightness level for the given backlight as a percent
 func get_brightness(backlight_path: String) -> float:
-	var cur_brightness := get_brightness_value(backlight_path)
-	var max_brightness := get_max_brightness_value(backlight_path)
-	if cur_brightness == -1 or max_brightness == -1:
-		return -1
-	return float(cur_brightness) / max_brightness
+	if brightness_provider == null:
+		return -1.0
+	return brightness_provider.get_brightness(backlight_path)
 
 
 ## Returns the current brightness value for the given backlight
 func get_brightness_value(backlight_path: String) -> int:
-	var brightness_file := "/".join([backlight_path, "brightness"])
-	var output := []
-	var code := OS.execute("cat", [brightness_file], output)
-	if code != OK:
-		logger.debug("Unable to get current brightness: " + output[0])
+	if brightness_provider == null:
 		return -1
-	var value := (output[0] as String).strip_edges()
-	if value.is_valid_int():
-		return value.to_int()
-	logger.debug("Brightness is not a valid number: " + value)
-	return -1
+	return brightness_provider.get_brightness_value(backlight_path)
 
 
 ## Returns the maximum brightness for the given backlight
 func get_max_brightness_value(backlight_path: String) -> int:
-	var max_brightness_file := "/".join([backlight_path, "max_brightness"])
-	var output := []
-	var code := OS.execute("cat", [max_brightness_file], output)
-	if code != OK:
-		logger.debug("Unable to get max brightness: " + output[0])
+	if brightness_provider == null:
 		return -1
-	var value := (output[0] as String).strip_edges()
-	if value.is_valid_int():
-		return value.to_int()
-	logger.debug("Max brightness is not a valid number: " + value)
-	return -1
+	return brightness_provider.get_max_brightness_value(backlight_path)
 
 
 ## Returns a list of all detected backlight devices
 func get_backlight_paths() -> PackedStringArray:
-	var backlights := PackedStringArray()
-	var backlight_dir := DirAccess.open(backlight_path)
-	var devices := backlight_dir.get_directories()
-	for device in devices:
-		backlights.append("/".join([backlight_path, device]))
-	logger.debug("Found backlights: " + str(backlights))
-	return backlights
+	if brightness_provider == null:
+		return PackedStringArray()
+	return brightness_provider.get_backlights()
 
 
-func _get_brightness_provider() -> BRIGHTNESS_PROVIDER:
-	if FileAccess.file_exists(steamos_write_bin):
+func _get_backlight_provider() -> BacklightProvider:
+	if FileAccess.file_exists("/usr/bin/steamos-polkit-helpers/steamos-priv-write"):
```

Comment:
Should we consider swapping the priority here for distros that have both? What's the benefits for OS.execute vs Command.execute_blocking?